datasets.py

```diff
@@ -117,6 +122,89 @@ def read_xml_content(xml_file):
+    def dist_calculator(word):
+        assert(word is not None)
+        # create list of supercats for easy indexing & ordering
+        supercat_list = list(supercategories_to_names.keys())
+        # score array for each supercat, for softmax calculation
+        score_arr = []
+
+        for supercat in (supercat_list):
+            word1 = nlp(supercategories_to_names[supercat])
+            word2 = nlp(word)
+            cur_score = word1.similarity(word2)
+            if cur_score is None:
+                score_arr.append(-1)
+                continue
+            else:
+                score_arr.append(cur_score)
+        score_arr = softmax(score_arr)
+        
+        return supercat_list[np.argmax(score_arr)], np.max(score_arr)
+    ######################################################################    
+    # this represents the result map from label to supercat
+    result_label_to_group_map = OrderedDict()
+    # (hr_label) is the human-readable value corresponding to the key (label)
+    for label, hr_label in labels_to_names.items():
+        if override_map and hr_label in override_map:
+            # skip if user is overriding with override_map
+            continue
+        # retrieve tuple result from dist_calculator
+        supercat_match = dist_calculator(hr_label)
+        result_label_to_group_map[label] = supercat_match
+        
+    # sort from least to most confident based on softmax scores
+    result_label_to_group_map = dict(sorted(result_label_to_group_map.items(), key=lambda item: item[1][1]))
+    
+    # remove the softmax scores from the dictionary
+    for k, v in result_label_to_group_map.items():
+        result_label_to_group_map[k] = v[0]
+    
+    # add the harded-coded override_map values if provided:
+    if override_map:
+        # initialize non-human-readable form of override_map
+        override_nonhuman_readable = {}
+        # invert the labels_to_names 
+        names_to_labels = dict((v, k) for k, v in labels_to_names.items())
+        # invert the supercategories_to_names
+        names_to_supercategories = dict((v, k) for k, v in supercategories_to_names.items())
+        # convert human-readable labels and supercats 
+        for k,v in override_map.items():
+            label = names_to_labels.get(k)
+            supercat = names_to_supercategories.get(v)
+            assert label in labels_to_names and supercat in supercategories_to_names, "override string not valid"
+            override_nonhuman_readable[label] = supercat
+        # add overriden dict to result dict 
+        result_label_to_group_map.update(override_nonhuman_readable)
+
+    # print mapping in human-readable form so user can adjust if necessary
+    print("Ranked from least to most confident, here are the \nlabel->supercategory mappings:\n(change as necessary using override_map)")
+    print("-------------------------------")
+    for k, v in result_label_to_group_map.items():
```

Comment:
does this print all of them? maybe just do the first 20 or something, and mention that in the comment

Reply:
done, printing the 20 least confident now